Userland/nc.cpp

```diff
@@ -0,0 +1,126 @@
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define max(a, b) ((a > b) ? a : b)
+
+int main(int argc, char** argv)
+{
+    if (argc < 3) {
+        fprintf(stderr, "Usage: nc <host> <port>\n");
+        return -1;
+    }
+
+    const char* addr_str = argv[1];
+    int port = atoi(argv[2]);
+
+    int fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (fd < 0) {
+        perror("socket");
+        return 1;
+    }
+
+    struct timeval timeout {
+        3, 0
+    };
+    int rc = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+    if (rc < 0) {
+        perror("setsockopt");
+        return 1;
+    }
+    rc = setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
```

Comment:
I don't think `SO_SNDTIMEO` is hooked up all the way, we should fix that too at some point :)